File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null)
            return;

        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode revHead = reverseList(slow.next);
        slow.next = null;

        mergeAlternate(head, revHead);
    }

    private ListNode reverseList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode newHead = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }

    private ListNode mergeAlternate(ListNode l1, ListNode l2) {
        ListNode head = l1;
        ListNode temp1, temp2;

        while (l1 != null && l2 != null) {
            temp1 = l1.next;
            temp2 = l2.next;

            l1.next = l2;
            if (temp1 == null) {
                break;
            }

            l2.next = temp1;

            l1 = temp1;
            l2 = temp2;
        }
        return head;
    }

}

File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function (head) {
    let slow = head, fast = head;
    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }
    const rHead = reverse(slow.next);
    slow.next = null;
    alternateMerge(head, rHead);
};

var reverse = (head) => {
    let prev = null;
    let curr = head;

    while (curr) {
        let next = curr.next;

        curr.next = prev;
        prev = curr;
        curr = next;
    }

    return prev;
}

var alternateMerge = (l1, l2) => {
    while (l1 && l2) {
        let next1 = l1.next;
        let next2 = l2.next;
        l1.next = l2;
        l2.next = next1;
        l1 = next1;
        l2 = next2;
    }
}